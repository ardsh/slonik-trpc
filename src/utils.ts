
export {
    rowToJson,
    rowsToArray,
    arrayFilter,
    booleanFilter,
    dateFilter,
    genericFilter,
    dateFilterType,
    arrayStringFilterType,
    invertFilter,
    arrayifyType,
} from './helpers/sqlUtils';

export {
    mergeFilters,
    makeFilter,
    createFilters
} from './core/queryFilter';

export {
    createResultParserInterceptor
} from './helpers/resultParserInterceptor';
