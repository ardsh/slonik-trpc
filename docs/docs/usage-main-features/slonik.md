---
sidebar_position: 20
---

# Slonik

This is an abbreviated guide to get started with slonik for the purposes of this tutorial, which is just a toy example.

You can also skip this step if you get started with the [minimal-example playground](https://stackblitz.com/github/ardsh/slonik-trpc/tree/main/examples/minimal-trpc)

You should refer to [slonik's comprehensive documentation](https://github.com/gajus/slonik#usage) for more advanced use cases.


## PostgreSQL

This is a good starting point for getting [free PostgreSQL database in the cloud](https://hasura.io/blog/top-postgresql-database-free-tier-solutions/)

Create a database in [Neon](https://neon.tech/), then export the DATABASE_URL by putting it in your .env file.

## Connecting

Install slonik

```bash
yarn add slonik
```

Create a file at `src/slonik.ts`:

```ts title="src/slonik.ts"
import { CommonQueryMethods, createPool, createTypeParserPreset, sql } from 'slonik';
import { createResultParserInterceptor } from "slonik-trpc/utils";

export const slonik = createPool(process.env.POSTGRES_DSN || process.env.DATABASE_URL, {
    interceptors: [createResultParserInterceptor()],
    typeParsers: [
    ...createTypeParserPreset().filter(
        (a) => a.name !== "timestamp" && a.name !== "timestamptz" && a.name !== "date"
    ), {
        name: "date",
        parse: (a) => !a || !Date.parse(a) ? a :
            new Date(a).toISOString().slice(0, 10),
    }, {
        name: "timestamptz",
        parse: (a) => !a || !Date.parse(a) ? a : new Date(a).toISOString(),
    }, {
        name: "timestamp",
        parse: (a) => !a || !Date.parse(a) ? a : new Date(a + "Z").toISOString(),
    }],
})

// If you're using ES modules with node 14+ you can use top-level await here
// export const db = await slonik;
export const db: CommonQueryMethods = new Proxy({} as never, {
    get(target, prop: keyof CommonQueryMethods) {
        return (...args: any[]) => {
            return pool.then((db) => {
                return Function.prototype.apply.apply(db[prop], [db, args]);
            });
        };
    },
});
```

We're adding specific type parsers for the timestamp/date types to make it easier by returning ISO strings, slonik returns int timestamps by default.

The DATABASE_URL env variable should take the form of `postgresql://user:password@host:port/database`

## Creating the database schema

If you'd like to create the database schema for this tutorial in SQL, simply create a `schema.ts` file.

```ts title="src/schema.ts"
import { db } from './slonik.ts';

export async function initializeDatabase(schema?: string) {
    if (schema) {
        await db.query(sql.unsafe`
            CREATE SCHEMA IF NOT EXISTS ${sql.identifier([schema])};
            SET search_path TO ${sql.identifier([schema])};
        `);
    }
    await db.query(sql.unsafe`
        DROP TABLE IF EXISTS users; DROP TABLE IF EXISTS posts;

        CREATE TABLE IF NOT EXISTS posts (
            id integer NOT NULL PRIMARY KEY,
            author_id text NOT NULL,
            title text NOT NULL,
            date date NOT NULL,
            content text NOT NULL DEFAULT '',
            created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
        );

        CREATE TABLE IF NOT EXISTS users (
            "id" text NOT NULL PRIMARY KEY,
            "first_name" text NOT NULL,
            "last_name" text NOT NULL,
            "email" text NOT NULL,
            "created_at" TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
        );`);
    await db.query(sql.unsafe`
        INSERT INTO posts
            (id, author_id, title, content, date)
        VALUES
            (1, 'z', 'aaa', 'This is a post', '2022-01-01'),
            (2, 'y', 'aaa', 'This is a post', '2022-02-01'),
            (3, 'x', 'bbb', 'This is a post', '2022-03-01'),
            (4, 'w', 'bbb', 'This is a post', '2022-04-01'),
            (5, 'v', 'ccc', 'This is a post', '2022-05-01'),
            (6, 'u', 'ccc', 'This is a post', '2022-06-01'),
            (7, 't', 'ddd', 'This is a post', '2022-07-01'),
            (8, 's', 'ddd', 'This is a post', '2022-08-01'),
            (9, 'r', 'eee', 'This is a post', '2022-09-01');

        INSERT INTO users
            (id, "first_name", "last_name", email)
        VALUES
            ('z', 'Haskell', 'Nguyen', 'haskell04@gmail.com'),
            ('y', 'Padberg', 'Fletcher', 'padberg.shawna@hotmail.com'),
            ('x', 'Neal', 'Phillips', 'nvandervort@collier.com'),
            ('w', 'Nolan', 'Muller', 'qnolan@yahoo.com'),
            ('v', 'Bob', 'Dean', 'acummerata@gmail.com'),
            ('u', 'Rebecca', 'Mercer', 'moore.rebeca@yahoo.com'),
            ('t', 'Katheryn', 'Ritter', 'katheryn89@hotmail.com'),
            ('s', 'Dulce', 'Espinoza', 'dulce23@gmail.com'),
            ('r', 'Paucek', 'Clayton', 'paucek.deangelo@hotmail.com');
    `);
}

initializeDatabase('playground');
```
