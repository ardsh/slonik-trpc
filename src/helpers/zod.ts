import { z, ZodError } from 'zod';
import { SchemaValidationError } from 'slonik';

export function notEmpty<TValue>(value: TValue | null | undefined | void): value is TValue {
    return value !== null && value !== undefined;
}

export const arrayifyType = <T extends z.ZodType>(type: T) =>
z.preprocess(
    (a) => (Array.isArray(a) ? a : [a].filter(notEmpty)),
    z.union([z.array(type), type])
);
