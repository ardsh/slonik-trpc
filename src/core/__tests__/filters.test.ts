import { sql } from 'slonik';
import { z } from 'zod';
import { createFilters, makeFilter, mergeFilters, arrayFilter, booleanFilter, dateFilter, dateFilterType, arrayifyType } from '../../utils';

const filtersOptions = createFilters()({
    id: arrayifyType(z.number()),
    uid: arrayifyType(z.string()),
    largeIds: z.boolean(),
    date: dateFilterType
}, {
    id: (text) => sql.fragment`"id" = ${text}`,
    date: async (date) => {
        await new Promise((resolve) => setTimeout(resolve, 1));
        return dateFilter(date, sql.fragment`"date"`);
    },
    // If true is specified, id must be greater than 5
    largeIds: (filter) => booleanFilter(filter, sql.fragment`"id" > 5`),
    uid: async (uids) => arrayFilter(uids, sql.fragment`"uid"`),
}, {
    postprocess(conditions) {
        conditions.push(sql.fragment`uid = 'x'`)
        return conditions;
    }
});
const getConditions = makeFilter(filtersOptions.interpreters, filtersOptions.options);

type Filter = Parameters<typeof getConditions>[0];
const filters = [] as [string, Filter][];

const addFilter = (text: string, filter: Filter) => {
    filters.push([text, filter]);
}

addFilter("Filters on date fields", {
    date: {
        _gt: "2022-02-03",
    }
});

addFilter("Filters on array string fields with a single string", ({
    uid: 'z',
}))

addFilter("Filters on array string fields with multiple", ({
    uid: ['y', 'x'],
}))

addFilter("AND Filters", ({
    AND: [{
        uid: ['y', 'x'],
    }, {
        id: 3,
    }]
}))

addFilter("AND Filters nested with OR", ({
    AND: [{
        OR: [{
            uid: ['y', 'x'],
        }, {
            largeIds: false,
        }]
    }, {
        id: 3,
    }]
}));

it.each(filters)(
    "%s", async (description, data) => {
        expect(await getConditions(data)).toMatchSnapshot();
    }
);


const otherFilters = createFilters()({
    isCheap: z.boolean(),
    isBeforeNow: z.boolean(),
}, {
    isCheap: (filter) => booleanFilter(filter, sql.fragment`"value" ILIKE '%cheap%'`),
    isBeforeNow: (filter) => booleanFilter(filter, sql.fragment`"date" < NOW()`),
}, {
    preprocess(filters) {
        if (filters) {
            filters.isCheap = true;
        }
        return filters;
    }
});

const anotherFilters = createFilters()({
    randomNumber: arrayifyType(z.number()),
    valueFilter: arrayifyType(z.string()),
}, {
    randomNumber: (text) => sql.fragment`"id" = ${sql.array(text as number[], 'numeric')}`,
    valueFilter: (value) => arrayFilter(value, sql.fragment`"value"`),
}, {
    postprocess(conditions) {
        conditions.push(sql.fragment`uid = 'anotherFilter'`);
        return conditions;
    }
});

const combinedFilter = mergeFilters([filtersOptions, otherFilters, anotherFilters]);

// const loader = makeQueryLoader({
//     query: sql.type(z.object({ id: z.number() }))`SELECT 2 AS "id"`,
//     filters: combinedFilter,
// });

const getCombinedConditions = makeFilter(createFilters()(combinedFilter.filters, combinedFilter.interpreters, combinedFilter.options).interpreters, combinedFilter.options);

type CombinedFilter = Parameters<typeof getCombinedConditions>[0];
const combined = filters.slice() as [string, CombinedFilter][];

const addCombinedFilter = (text: string, filter: CombinedFilter) => {
    combined.push([text, filter]);
}

addCombinedFilter("AND Filters deeply nested with OR", ({
    AND: [{
        OR: [{
            uid: ['y', 'x'],
            NOT: {
                isCheap: false,
            }
        }, {
            isBeforeNow: false,
        }]
    }, {
        randomNumber: [2, 1],
    }]
}));

addCombinedFilter("AND with empty OR array", ({
    AND: [{
        OR: [],
        randomNumber: [3, 4],
    }]
}));

addCombinedFilter("OR with empty AND array", ({
    OR: [{
        AND: [],
        valueFilter: ['test'],
    }]
}));

addCombinedFilter("Empty AND and OR arrays", ({
    AND: [],
    OR: [],
    isCheap: true,
}));

it.each(combined)(
    "%s", async (description, data) => {
        expect(await getCombinedConditions(data)).toMatchSnapshot();
    }
);
