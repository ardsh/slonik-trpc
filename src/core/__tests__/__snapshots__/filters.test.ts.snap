// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AND Filters 1`] = `
{
  "sql": "((("uid" = ANY($1::"text"[]))) AND (("id" = $2))) AND (uid = 'x')",
  "type": "SLONIK_TOKEN_FRAGMENT",
  "values": [
    [
      "y",
      "x",
    ],
    3,
  ],
}
`;

exports[`AND Filters 2`] = `
{
  "sql": "("value" ILIKE '%cheap%') AND ((("uid" = ANY($1::"text"[]))) AND (("id" = $2))) AND (uid = 'x') AND (uid = 'anotherFilter')",
  "type": "SLONIK_TOKEN_FRAGMENT",
  "values": [
    [
      "y",
      "x",
    ],
    3,
  ],
}
`;

exports[`AND Filters deeply nested with OR 1`] = `
{
  "sql": "("value" ILIKE '%cheap%') AND ((((("uid" = ANY($1::"text"[])) AND ((("value" ILIKE '%cheap%')))) OR ((NOT ( "date" < NOW() ))))) AND (("id" = $2::"numeric"[]))) AND (uid = 'x') AND (uid = 'anotherFilter')",
  "type": "SLONIK_TOKEN_FRAGMENT",
  "values": [
    [
      "y",
      "x",
    ],
    [
      2,
      1,
    ],
  ],
}
`;

exports[`AND Filters nested with OR 1`] = `
{
  "sql": "((((("uid" = ANY($1::"text"[]))) OR ((NOT ( "id" > 5 ))))) AND (("id" = $2))) AND (uid = 'x')",
  "type": "SLONIK_TOKEN_FRAGMENT",
  "values": [
    [
      "y",
      "x",
    ],
    3,
  ],
}
`;

exports[`AND Filters nested with OR 2`] = `
{
  "sql": "("value" ILIKE '%cheap%') AND ((((("uid" = ANY($1::"text"[]))) OR ((NOT ( "id" > 5 ))))) AND (("id" = $2))) AND (uid = 'x') AND (uid = 'anotherFilter')",
  "type": "SLONIK_TOKEN_FRAGMENT",
  "values": [
    [
      "y",
      "x",
    ],
    3,
  ],
}
`;

exports[`Filters on array string fields with a single string 1`] = `
{
  "sql": "("uid" = ANY($1::"text"[])) AND (uid = 'x')",
  "type": "SLONIK_TOKEN_FRAGMENT",
  "values": [
    [
      "z",
    ],
  ],
}
`;

exports[`Filters on array string fields with a single string 2`] = `
{
  "sql": "("uid" = ANY($1::"text"[])) AND ("value" ILIKE '%cheap%') AND (uid = 'x') AND (uid = 'anotherFilter')",
  "type": "SLONIK_TOKEN_FRAGMENT",
  "values": [
    [
      "z",
    ],
  ],
}
`;

exports[`Filters on array string fields with multiple 1`] = `
{
  "sql": "("uid" = ANY($1::"text"[])) AND (uid = 'x')",
  "type": "SLONIK_TOKEN_FRAGMENT",
  "values": [
    [
      "y",
      "x",
    ],
  ],
}
`;

exports[`Filters on array string fields with multiple 2`] = `
{
  "sql": "("uid" = ANY($1::"text"[])) AND ("value" ILIKE '%cheap%') AND (uid = 'x') AND (uid = 'anotherFilter')",
  "type": "SLONIK_TOKEN_FRAGMENT",
  "values": [
    [
      "y",
      "x",
    ],
  ],
}
`;

exports[`Filters on date fields 1`] = `
{
  "sql": "(("date" > $1)) AND (uid = 'x')",
  "type": "SLONIK_TOKEN_FRAGMENT",
  "values": [
    "2022-02-03",
  ],
}
`;

exports[`Filters on date fields 2`] = `
{
  "sql": "(("date" > $1)) AND ("value" ILIKE '%cheap%') AND (uid = 'x') AND (uid = 'anotherFilter')",
  "type": "SLONIK_TOKEN_FRAGMENT",
  "values": [
    "2022-02-03",
  ],
}
`;
